### ISSUE TITLE
[BUG] Playwright seems to get stuck when using shard option and last test is skipped

### ISSUE THREAD

**Issue by jfpimentel** [microsoft/playwright#21226]
**Context:**
- Playwright Version: 1.31.1
- Operating System: Mac
- Node.js version: 18.13.0
- Browser: Chromium

**Describe the bug**

I've just started using the `--shard=x/y` option as described [in the docs](https://playwright.dev/docs/test-parallel#shard-tests-between-multiple-machines), but it does not seem to work well when some of the tests are skipped. Please see the pictures below of shards 1 and 2:
- in `--shard=1/2`, the last 3 tests are skipped and Playwright seems to get stuck and never finishes
- in `--shard=2/2`, all of the tests run and succeed, and the process ends as expected

<img width="801" alt="Screenshot 2023-02-27 at 11 55 12" src="https://user-images.githubusercontent.com/30432768/221561508-e23a81bc-6315-4d2f-8f7f-7220e7413068.png">

<img width="877" alt="Screenshot 2023-02-27 at 11 58 02" src="https://user-images.githubusercontent.com/30432768/221560594-6b79630d-0a0c-4dce-87c2-662e7fcc826f.png">

**Comment by EelcoLos** [comment]
Can confirm. We had this same issue, which occured due to dependabot updating the `@playwright/test` dependency.
Another thing that the amount of tests per shard were incorrect. The shard with skipped tests had the majority of tests (13 of 26 tests in total).
Our temp solution was to revert the package back to 1.30.0 and (for now) adding the package to dependabots ignore list

**Comment by pavelfeldman** [comment]
@jfpimentel I can repro, we'll fix it in a patch release shortly.

@EelcoLos the sharding granularity has been changed between 1.30 and 1.31. We used to shard tests and now we shard entire files. This improved users' expectations and made the use of beforeAll hooks more efficient, but if you have small number of files and unbalanced numbers of tests in these files, it'll hurt you. The solution would be to balance the files.
